datafusion/functions/src/core/nullif.rs

```diff
@@ -122,8 +121,13 @@ fn nullif_func(args: &[ColumnarValue]) -> Result<ColumnarValue> {
             Ok(ColumnarValue::Array(array))
         }
         (ColumnarValue::Scalar(lhs), ColumnarValue::Array(rhs)) => {
-            let lhs = lhs.to_array_of_size(rhs.len())?;
-            let array = nullif(&lhs, &eq(&lhs, &rhs)?)?;
+            let lhs_s = lhs.to_scalar()?;
+            let lhs_a = lhs.to_array_of_size(rhs.len())?;
```

Comment:
Maybe we could update the arrow-rs nullif kernel to have a special case for a constant (arrow-rs calls this idea "Datum" rather than ScalarValue):

https://docs.rs/arrow/latest/arrow/array/trait.Datum.html

But you can get the `Datum` like this:https://docs.rs/datafusion/latest/datafusion/common/enum.ScalarValue.html#method.to_scalar

Reply:
Specializing arrow-rs's nullif kernel for a constant would be the absolute best. But if others can reproduce the small % perf increase of this PR, then perhaps it can be merged in isolation?

Reply:
I think this is a good improvement regardless as it makes the DataFusion code follow the pattern so it can take advantage of the `Datum` special case if/when it is implemented.

Any chance you have a moment to file a ticket upstream in arrow-rs? If not, I will do so